### ISSUE TITLE
dolly-v2-3b with thunder_inductor_cat_cudnn fails with KeyError: \'t5905\'

### ISSUE THREAD

**Issue by wprazuch** [Lightning-AI/lightning-thunder#665]
## 🐛 Bug

There is a key error error when running models:
- dolly-v2-3b

for fsdp zero2/zero3 two node setup (2 nodes per 2 gpus each).

### To Reproduce

(example for one model)

```
mkdir -p output
docker run --pull=always --gpus all --ipc=host --ulimit memlock=-1 --ulimit stack=67108864  -v $PWD/output:/output -it INTERNAL_IMAGE:pjnl-20240621
```
Run in the container:

```
torchrun --nnodes=2 --nproc-per-node=8 /opt/pytorch/lightning-thunder/thunder/benchmarks/benchmark_litgpt.py --model_name dolly-v2-3b --compile thunder_inductor_cat_cudnn --distributed_mode fsdp --shard_mode zero2 
```

### Expected behavior

The model should run or we should get OOM error.

### Environment

As in Docker image.

### Additional context

Traceback:
```
[rank0]: Traceback (most recent call last):
[rank0]:   File "/workspace/lightning-thunder/thunder/benchmarks/benchmark_litgpt.py", line 639, in <module>
[rank0]:     CLI(benchmark_main)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/jsonargparse/_cli.py", line 96, in CLI
[rank0]:     return _run_component(components, cfg_init)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/jsonargparse/_cli.py", line 196, in _run_component
[rank0]:     return component(**cfg)
[rank0]:   File "/workspace/lightning-thunder/thunder/benchmarks/benchmark_litgpt.py", line 584, in benchmark_main
[rank0]:     benchmark.train()
[rank0]:   File "/workspace/lightning-thunder/thunder/benchmarks/benchmark_litgpt.py", line 491, in train
[rank0]:     loss.backward()
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/_tensor.py", line 522, in backward
[rank0]:     torch.autograd.backward(
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/autograd/__init__.py", line 288, in backward
[rank0]:     _engine_run_backward(
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/autograd/graph.py", line 768, in _engine_run_backward
[rank0]:     return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/autograd/function.py", line 306, in apply
[rank0]:     return user_fn(self, *args)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/autograd/function.py", line 599, in wrapper
[rank0]:     outputs = fn(ctx, *args)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/executors/torch_autograd.py", line 96, in backward
[rank0]:     grads = ctx.compiled_backward([saved_tensors_list, ctx.saved_other], args)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/amp/autocast_mode.py", line 43, in decorate_autocast
[rank0]:     return func(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/amp/autocast_mode.py", line 43, in decorate_autocast
[rank0]:     return func(*args, **kwargs)
[rank0]:   File "thunder.backward_fn_269", line 514, in backward_fn
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/executors/nvfuserex_impl.py", line 402, in __call__
[rank0]:     fd = self.get_fd(to_descriptors(args))
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/executors/nvfuserex_impl.py", line 512, in get_fd
[rank0]:     return create_fd(bsyms, input_descriptors, sorted_unique_inputs, sorted_unique_outputs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/executors/nvfuserex_impl.py", line 274, in create_fd
[rank0]:     translate_bound_symbol(bsym)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/executors/nvfuserex_impl.py", line 264, in translate_bound_symbol
[rank0]:     nvresults = translator(*bsym.args, **bsym.kwargs, fd=fd, lc_to_nv_map=lc_to_nv_map)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/executors/nvfuserex_impl.py", line 1846, in sub
[rank0]:     nva = getnv(a, fd, lc_to_nv_map)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/executors/nvfuserex_impl.py", line 116, in getnv
[rank0]:     return lc_to_nv_map[x]
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/utils.py", line 939, in __getitem__
[rank0]:     return self._dict[key_]
[rank0]: KeyError: 't5905'
```

**Comment by wprazuch** [comment]
@IvanYashchuk this is related to the [past issue](https://github.com/Lightning-AI/lightning-thunder/issues/588).
Only this KeyError persists in the new container version :)
All the best,
WP

**Comment by IvanYashchuk** [comment]
Running `python thunder/benchmarks/benchmark_litgpt.py --model_name dolly-v2-3b --compile thunder_inductor_cat_cudnn --n_layers=1` doesn't reproduce the problem. Need to check on DGX when it's available.

**Comment by kiya00** [comment]
seems already fixed, I tried it on today's container on H100:
```
root@c6f2bbeb93de:/opt/pytorch/lightning-thunder# torchrun --nproc-per-node=8 /opt/pytorch/lightning-thunder/thunder/benchmarks/benchmark_litgpt.py --model_name dolly-v2-3b --compile thunder_inductor_cat_cudnn --distributed_mode fsdp --shard_mode zero2
Model name: dolly-v2-3b
Seq Length: 2048
Micro BS: 1
Global BS: 8
Number of Layers: 32
Number of parameters: 0.35B
Distributed Mode: fsdp
Sharding Mode: zero2
Bucketing: none
Compiler: thunder_inductor_cat_cudnn
Average iter time: 113.18 ms
Memory used: 17.74 GB
Tokens/s: 144833.85
Tokens/s/GPU: 18104.23
TFLOP/s: 2590.36
```

**Comment by IvanYashchuk** [comment]
Thank you, Yan, for checking this!
The reported error is also not reproducible with the provided tag `pjnl-20240621` on a single node. Is the tag correct?

@wprazuch, in the issue description, you write that the error is seen when running on two nodes, but the reproducing command doesn't have the `--nnodes=` option. I don't know how the two-node run could differ from one node, but anything is possible, we'll check.

**Comment by wprazuch** [comment]
@IvanYashchuk Yes, you are right about `--nnodes` option, it should be set to 2. The issue only persists for multi-node setup, because we got it for:
- nodes = 2
- gpu_per_node = 8
- fsdp
- zero2 and zero3

I forgot to add the option in reproduction command - unfortunately, due to manual work needed for adjusting commands for this repo I missed that. I adjusted the command in the issue description - sorry for the confusion.
